sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -633,6 +634,13 @@ private synchronized void initKinesisProducer() {
         producer = spec.getAWSClientsProvider().createKinesisProducer(config);
       }
 
+      private void readObject(ObjectInputStream is) throws IOException, ClassNotFoundException {
+        is.defaultReadObject();
+        if (producer == null) {
```

Comment:
Don't you think that it would be better to do this check in `startBundle()`? 
Also, it would be great to have a regression test for this case.

Reply:
Thank you for looking into this. Do you mean add this check in startBundle() and then call initKinesisProducer directly in readObject?  And for the regression test, may I get a pointer to the class where that would be appropriate to add?

Reply:
I meant just to add this check:
```
if (producer == null) {
  initKinesisProducer();
}
```
directly into `KinesisWriterFn.startBundle()`. Beam will guarantee that the whole bundle will be processed on the same worker and the object won't be serialised/deserialised until 
 this bundle will be finished. Though, I think `readObject` should do the same job. Wdyt?

Regarding the test, I believe that all Kinesis.Write tests are in `KinesisMockWriteTest`, but now I started to doubt if it's possible to test such regression in unit tests. Wdyt?

Reply:
I will add the check in startBundle and verify that it does the same thing as readObject. Yeah, I think this wouldn't be caught in a unit test.

Reply:
To avoid code duplication, I suggest to move this `if (producer == null){...}` block into `initKinesisProducer()`, so it will be lazy producer initialisation.

Reply:
@aromanenko-dev - Please look at the changes one more time. Thank you.

Reply:
Thanks, though I see that the test  `org.apache.beam.sdk.io.kinesis.KinesisMockWriteTest.testWriteFailed` started failing (see Java PreCommit check)